Firestore/src/WriteBatch.php

```diff
@@ -712,23 +715,47 @@ private function filterFields(array $fields, array $inputPaths = [])
 
                         if ($illegalNestedDelete) {
                             throw new \InvalidArgumentException(sprintf(
-                                '%s::deleteField() values cannot be nested.',
-                                FieldValue::class
+                                '%s::deleteField() values cannot be nested. ' .
+                                'Invalid value found at %s.',
+                                FieldValue::class,
+                                $currentPath->pathString()
                             ));
                         }
 
+                        // Values of type `DocumentTransformInterface` cannot contain nested transforms.
+                        if ($value instanceof DocumentTransformInterface) {
+                            $args = $value->args();
+                            if (is_array($args) && !$this->isAssoc($args)) {
+                                foreach ($args as $arg) {
+                                    if ($arg instanceof DocumentTransformInterface) {
+                                        throw new \InvalidArgumentException(sprintf(
```

Comment:
Note that this can be nested:

```
docRef.update("a", FieldValue.arrayUnion("b", "{c:FieldValue.arrayUnion("d"))
```

This should be rejected as well.

Reply:
FieldValue arguments are not parsed for nested FieldValues, so the case you cite is already rejected.

```php
$doc->set([
    'a' => FieldValue::arrayUnion([
        ['a' => FieldValue::arrayUnion([])]
    ])
]);
```

```
RuntimeException: Object of type Google\Cloud\Firestore\FieldValue\ArrayUnionValue cannot be encoded to a Firestore value type.
```

Reply:
Sounds good. The Node client dropped them silently, which was strictly worse. Thanks!